File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;

const int inf = 1e9;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, m; cin >> n >> m;
    vector<int> dp(n+1, inf);
    vector<vector<pii>> e(n+1);
    for(int i = 0; i < m; i++){
        int u, v, w; cin >> u >> v >> w;
        e[u].push_back({v, w});
    }
    int start, end; cin >> start >> end;
    priority_queue<pii> pq; pq.push({0, start});
    dp[start] = 0;
    while(pq.size()){
        auto [cost, cur] = pq.top(); pq.pop();
        if(dp[cur] < cost) continue;
        for(auto [nxt, ncost] : e[cur]){
            if(dp[nxt] > cost+ncost){
                dp[nxt] = cost+ncost;
                pq.push({dp[nxt], nxt});
            }
        }
    }
    cout << dp[end];
    return 0;
} 

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

from math import inf
import heapq

def main():
    n = int(input())
    m = int(input())
    dp = [inf]*(n+1)
    e = [[] for _ in range(n+1)]
    for _ in range(m):
        u, v, w = map(int, input().split())
        e[u].append((v, w))
    start, end = map(int, input().split())
    dp[start] = 0
    hq = []
    heapq.heappush(hq, (0, start))
    while hq:
        cost, cur = heapq.heappop(hq)
        if dp[cur] < cost: continue
        for nxt, ncost in e[cur]:
            if dp[nxt] > cost+ncost: 
                dp[nxt] = cost+ncost
                heapq.heappush(hq, (dp[nxt], nxt))
    print(dp[end])
if __name__ == '__main__':
    main()